marconi-chain-index/src/Marconi/ChainIndex/Indexers/EpochStakepoolSize.hs

```diff
@@ -1,187 +1,502 @@
+                     FROM epoch_spd
+                     WHERE epochNo = ?
+                  |] (SQL.Only epochNo)
+              pure $ SPDByEpochNoResult res
+
+    queryStorage _ _ EpochSPDHandle {} (LedgerStateAtPointQuery C.ChainPointAtGenesis) = do
+        pure $ LedgerStateAtPointResult Nothing
+    queryStorage
+            _
+            events
+            (EpochSPDHandle _ ledgerStateDirPath _)
+            (LedgerStateAtPointQuery (C.ChainPoint slotNo _)) = do
+        case List.find (\e -> epochSPDEventSlotNo e == slotNo) (toList events) of
+            Nothing -> do
+                ledgerStateFilePaths <- listDirectory ledgerStateDirPath
+                let ledgerStateFilePath =
+                        List.find
+                            (\fp -> fmap (\(_, sn, _, _) -> sn)
+                                         (chainTipsFromLedgerStateFilePath fp) == Just slotNo
+                            )
+                            ledgerStateFilePaths
+                case ledgerStateFilePath of
+                  Nothing -> pure $ LedgerStateAtPointResult Nothing
+                  Just fp -> do
+                      ledgerStateBs <- BS.readFile $ ledgerStateDirPath </> fp
+                      let ledgerState =
+                              either
+                                (const Nothing)
+                                (Just . snd)
+                                $ CBOR.deserialiseFromBytes decodeLedgerState ledgerStateBs
+                      pure $ LedgerStateAtPointResult ledgerState
+            Just event -> pure $ LedgerStateAtPointResult $ epochSPDEventLedgerState event
+
+instance Rewindable EpochSPDHandle where
+    rewindStorage
+        :: C.ChainPoint
+        -> EpochSPDHandle
+        -> IO (Maybe EpochSPDHandle)
+    rewindStorage C.ChainPointAtGenesis h@(EpochSPDHandle c ledgerStateDirPath _) = do
+        SQL.execute_ c "DELETE FROM epoch_spd"
+
+        ledgerStateFilePaths <- listDirectory ledgerStateDirPath
+        forM_ ledgerStateFilePaths (\f -> removeFile $ ledgerStateDirPath </> f)
+        pure $ Just h
+    rewindStorage (C.ChainPoint sn _) h@(EpochSPDHandle c ledgerStateDirPath _) = do
+        SQL.execute c "DELETE FROM epoch_spd WHERE slotNo > ?" (SQL.Only sn)
+
+        ledgerStateFilePaths <- listDirectory ledgerStateDirPath
+        forM_ ledgerStateFilePaths $ \fp -> do
+            case chainTipsFromLedgerStateFilePath fp of
+              Nothing                              -> pure ()
+              Just (_, slotNo, _, _) | slotNo > sn -> removeFile $ ledgerStateDirPath </> fp
+              Just _                               -> pure ()
+
+        pure $ Just h
+
+instance Resumable EpochSPDHandle where
+    resumeFromStorage
+        :: EpochSPDHandle
+        -> IO [C.ChainPoint]
+    resumeFromStorage (EpochSPDHandle c ledgerStateDirPath _) = do
+        epochSPDChainPoints :: Set C.ChainPoint <- Set.fromList . fmap (uncurry C.ChainPoint) <$>
```

Comment:
Since we can search in sqlite then it might be better to load the available chain points from the ledger states first, and then search for these in sqlite. Currently it loads *all* chain points from database which is large.

Reply:
Spent an hour tackling this. Seems like there is no clear way to do something like `WHERE slotNo IN (?) AND blockHeaderHash = (?)` where the `?` is a list of elements. 

Closest I've managed is:

```
    resumeFromStorage
        :: EpochSPDHandle
        -> IO [C.ChainPoint]
    resumeFromStorage (EpochSPDHandle c ledgerStateDirPath _) = do
        ledgerStateFilepaths <- listDirectory ledgerStateDirPath
        let ledgerStateChainPoints =
                fmap (\(_, sn, bhh, _) -> (sn, bhh))
                $ mapMaybe chainTipsFromLedgerStateFilePath ledgerStateFilepaths

        let ledgerStateChainPointsValues =
                fmap (\(slotNo, bhh) -> (slotNo, bhh)) ledgerStateChainPointsledgerStateChainPointsValues
        epochSPDChainPoints :: [C.ChainPoint] <- fmap (uncurry C.ChainPoint) <$>
                SQL.query c
                    [r|SELECT slotNo, blockHeaderHash
                       FROM epoch_spd
                       WHERE slotNo IN (?) AND blockHeaderHash in (?)|]
                    ( Text.intercalate ","
                        $ fmap (\(C.SlotNo slotNo, _) -> Text.pack $ show slotNo) ledgerStateChainPointsValues
                    , Text.intercalate ","
                        $ fmap (\(_, bhh) -> "\"" <> C.serialiseToRawBytesHexText bhh <> "\"") ledgerStateChainPointsValues
                    )
```

However, that seems hacky. Also, I can't match the blockHeaderHash in the table by using the WHERE clause. I thought that was because the `blockHeaderHash` SQL field was a blob. Change it to `TEXT` but doesn't change anything: can't query by blockHeaderHash.

Any pointers or something that I've missed?

Reply:
Tried this implementation and seems to work: 56a74cfd1999ed2f23206b865b2e558e1d6f29ef
What do you think?

Reply:
Seems `in` is not supported :( https://github.com/nurpax/sqlite-simple/issues/91, regarding 56a74cfd1999ed2f23206b865b2e558e1d6f29ef, returning early would be better, and I guess the other idea would be to splice the `where in ...` string together manually.